security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/UserMappingOAuth2Service.scala

```diff
@@ -4,25 +4,41 @@ import io.circe.Decoder
 import pl.touk.nussknacker.ui.security.api.AuthenticatedUser
 import sttp.client3.SttpBackend
 
-import java.time.Instant
 import scala.concurrent.{ExecutionContext, Future}
 
 class UserMappingOAuth2Service[UserInfoData: Decoder, AuthorizationData <: OAuth2AuthorizationData: Decoder](
     delegate: OAuth2Service[UserInfoData, AuthorizationData],
-    loggedUserFunction: UserInfoData => AuthenticatedUser
+    loggedUserFunction: LoggedUserFunctionParameters[UserInfoData] => Future[AuthenticatedUser]
 )(implicit ec: ExecutionContext, backend: SttpBackend[Future, Any])
     extends OAuth2Service[AuthenticatedUser, AuthorizationData] {
 
   def obtainAuthorizationAndUserInfo(
       authorizationCode: String,
       redirectUri: String
-  ): Future[(AuthorizationData, AuthenticatedUser)] =
-    delegate.obtainAuthorizationAndUserInfo(authorizationCode, redirectUri).map { case (authorization, userInfo) =>
-      (authorization, loggedUserFunction(userInfo))
-    }
-
-  def checkAuthorizationAndObtainUserinfo(accessToken: String): Future[(AuthenticatedUser, Option[Instant])] =
-    delegate.checkAuthorizationAndObtainUserinfo(accessToken).map { case (userInfo, expiration) =>
-      (loggedUserFunction(userInfo), expiration)
-    }
+  ): Future[(AuthorizationData, AuthenticatedUser)] = {
+    for {
+      (authorization, userInfo) <- delegate.obtainAuthorizationAndUserInfo(authorizationCode, redirectUri)
+      loggedUser <- loggedUserFunction(
+        LoggedUserFunctionParameters(IntrospectedAccessTokenData.empty, () => Future.successful(userInfo))
+      )
+    } yield (authorization, loggedUser)
+  }
+
+  override def introspectAccessToken(accessToken: String): Future[IntrospectedAccessTokenData] =
+    delegate.introspectAccessToken(accessToken)
+
+  override def obtainUserInfo(
+      accessToken: String,
+      accessTokenData: IntrospectedAccessTokenData
+  ): Future[AuthenticatedUser] = {
+    loggedUserFunction(
+      LoggedUserFunctionParameters(accessTokenData, () => delegate.obtainUserInfo(accessToken, accessTokenData))
+    )
+  }
+
 }
+
+final case class LoggedUserFunctionParameters[UserInfoData](
```

Comment:
1. we have a function in it, so maybe just `final class` (with adding `val`s obviously)?
2. BTW, the class doesn't look "natural". Maybe we don't need it? ;) Maybe we won't need it if we have a trait instead of `loggedUserFunction: LoggedUserFunctionParameters[UserInfoData] => Future[AuthenticatedUser]`? WDYT?

Reply:
:+1:  In fact, we already had this trait - it was `OAuth2Profile`. I renamed it to `AuthenticationStrategy` and extracted `OAuth2Configuration` parameter to constructors which allowed me to remove this class.